tests/sharding/mainchain_handler/test_vmc_handler.py

```diff
@@ -0,0 +1,303 @@
+import logging
+import pytest
+
+import rlp
+
+from web3.exceptions import (
+    BadFunctionCallOutput,
+)
+
+from eth_tester.exceptions import (
+    ValidationError,
+)
+
+from eth_tester.backends.pyevm.main import (
+    get_default_account_keys,
+)
+
+from eth_utils import (
+    keccak,
+    to_canonical_address,
+    to_checksum_address,
+)
+
+from evm.utils.address import (
+    generate_contract_address,
+)
+from evm.utils.hexadecimal import (
+    encode_hex,
+)
+
+from evm.vm.forks.spurious_dragon.transactions import (
+    SpuriousDragonTransaction,
+)
+
+from evm.chains.sharding.mainchain_handler import (
+    vmc_utils,
+)
+
+from evm.chains.sharding.mainchain_handler.config import (
+    PERIOD_LENGTH,
+    TX_GAS,
+)
+
+from evm.chains.sharding.mainchain_handler.vmc_handler import (
+    VMC,
+)
+
+from evm.chains.sharding.mainchain_handler.vmc_utils import (
+    create_vmc_tx,
+    get_contract_address_from_contract_tx,
+    get_vmc_json,
+)
+
+from tests.sharding.mainchain_handler.fixtures import (  # noqa: F401
+    mainchain_handler,
+)
+
+PASSPHRASE = '123'
+ZERO_ADDR = b'\x00' * 20
+# for testing we set it to 5, 25 or 2500 originally
+SHUFFLING_CYCLE_LENGTH = 5
+
+test_keys = get_default_account_keys()
+
+logger = logging.getLogger('evm.chain.sharding.mainchain_handler.VMCHandler')
+
+
+def is_vmc_deployed(vmc_handler, chain_handler):
+    return (
+        # TODO: the following line should be uncommented when `get_code` is implemented in
+        #       `eth_tester`
+        # chain_handler.get_code(vmc_handler.address) != b'' and
+        chain_handler.get_nonce(vmc_handler.sender_addr) != 0
+    )
+
+
+def do_withdraw(vmc_handler, chain_handler, validator_index):
+    assert validator_index < len(test_keys)
+    privkey = test_keys[validator_index]
+    sender_addr = privkey.public_key.to_canonical_address()
+    signature = vmc_utils.sign(vmc_utils.WITHDRAW_HASH, privkey)
+    vmc_handler.withdraw(validator_index, signature, sender_addr)
+    chain_handler.mine(1)
+
+
+def deploy_valcode_and_deposit(vmc_handler, chain_handler, key):
+    """
+    Deploy validation code of and with the key, and do deposit
+
+    :param key: Key object
```

Comment:
Assuming this means `PrivateKey` object?

Reply:
Yep it is. I will change it. Thank you:)